pin/set.go

```diff
@@ -0,0 +1,327 @@
+package pin
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"sort"
+	"unsafe"
+
+	"github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/gogo/protobuf/proto"
+	"github.com/ipfs/go-ipfs/merkledag"
+	"github.com/ipfs/go-ipfs/pin/internal/pb"
+	"github.com/ipfs/go-ipfs/util"
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultFanout = 256
+	maxItems      = 8192
+)
+
+func randomSeed() (uint32, error) {
+	var buf [4]byte
+	if _, err := rand.Read(buf[:]); err != nil {
+		return 0, err
+	}
+	return binary.LittleEndian.Uint32(buf[:]), nil
+}
+
+func hash(seed uint32, k util.Key) uint32 {
+	var buf [4]byte
+	binary.LittleEndian.PutUint32(buf[:], seed)
+	h := fnv.New32a()
+	_, _ = h.Write(buf[:])
+	_, _ = io.WriteString(h, string(k))
+	return h.Sum32()
+}
+
+type itemIterator func() (k util.Key, data []byte, ok bool)
+
+type keyObserver func(util.Key)
+
+type refcount uint8
+
+func (r refcount) Bytes() []byte {
+	// refcount size can change in later versions; this may need
+	// encoding/binary
+	return []byte{byte(r)}
+}
+
+type sortByHash struct {
+	links []*merkledag.Link
+	data  []byte
+}
+
+func (s sortByHash) Len() int {
+	return len(s.links)
+}
+
+func (s sortByHash) Less(a, b int) bool {
+	return bytes.Compare(s.links[a].Hash, s.links[b].Hash) == -1
+}
+
+func (s sortByHash) Swap(a, b int) {
+	s.links[a], s.links[b] = s.links[b], s.links[a]
+	if len(s.data) != 0 {
+		const n = int(unsafe.Sizeof(refcount(0)))
+		tmp := make([]byte, n)
+		copy(tmp, s.data[a:a+n])
+		copy(s.data[a:a+n], s.data[b:b+n])
+		copy(s.data[b:b+n], tmp)
+	}
+}
+
+func storeItems(ctx context.Context, dag merkledag.DAGService, estimatedLen uint64, iter itemIterator, internalKeys keyObserver) (*merkledag.Node, error) {
```

Comment:
Yes, it's a mess, but it's not intended to stay permanently. If and when the API is ready for doing incremental changes, instead of this load and save dance, it'll look very different. So I'm not inclined to spend time of polishing it, I'd rather start evolving the pinning API.